test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,327 @@
+	if rc != 0 {
+		t.Errorf("Failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, ifName, stderr)
+	}
+
+	vm.ifName = strings.TrimSpace(ifName)
+	cmd = fmt.Sprintf("powershell 'Get-NetIPAddress -AddressFamily IPv4 -ifIndex %s| select IPAddress| ft -HideTableHeaders'", vm.ifIndex)
+	rc, ifIP, stderr, err := data.RunCommandOnNode(nodeName, cmd)
+	require.NoError(t, err, "Failed to run command <%s> on VM %s, err %v", cmd, nodeName, err)
+	if rc != 0 {
+		t.Errorf("Failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, ifIP, stderr)
+	}
+	vm.ip = strings.TrimSpace(ifIP)
+	return vm
+
+}
+
+func startAntreaAgent(t *testing.T, data *TestData, vm vmInfo) {
+	t.Logf("Starting antrea-agent on VM: %s", vm.nodeName)
+	var cmd string
+	if vm.osType == "Windows" {
+		cmd = "nssm start antrea-agent"
+	} else {
+		cmd = "sudo systemctl start antrea-agent"
+	}
+	rc, stdout, stderr, err := data.RunCommandOnNode(vm.nodeName, cmd)
+	require.NoError(t, err, "Failed to run command <%s> on VM %s, err %v", cmd, vm.nodeName, err)
+	if rc != 0 {
+		t.Errorf("Failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, stdout, stderr)
+	}
+}
+
+func stopAntreaAgent(t *testing.T, data *TestData, vm vmInfo) {
+	t.Logf("Stopping antrea-agent on VM: %s", vm.nodeName)
+	var cmd string
+	if vm.osType == "Windows" {
+		cmd = "nssm stop antrea-agent"
+	} else {
+		cmd = "sudo systemctl stop antrea-agent"
+	}
+	rc, stdout, stderr, err := data.RunCommandOnNode(vm.nodeName, cmd)
+	require.NoError(t, err, "Failed to run command <%s> on VM %s, err %v", cmd, vm.nodeName, err)
+	if rc != 0 {
+		t.Errorf("Failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, stdout, stderr)
+	}
+}
+
+func verifyInterfaceIsInOVS(t *testing.T, data *TestData, vm vmInfo) (found bool, err error) {
+	var cmd string
+	if vm.osType == "Windows" {
+		cmd = fmt.Sprintf("ovs-vsctl --column=name list port '%s'", vm.ifName)
+	} else {
+		cmd = fmt.Sprintf("sudo ovs-vsctl --column=name list port %s", vm.ifName)
+	}
+	rc, stdout, stderr, err := data.RunCommandOnNode(vm.nodeName, cmd)
+	if err != nil {
+		return false, fmt.Errorf("failed to run command <%s> on VM %s, err %v", cmd, vm.nodeName, err)
+	}
+	if rc != 0 {
+		return false, fmt.Errorf("failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, stdout, stderr)
+	}
+
+	if strings.Contains(stdout, "no row") {
+		return false, fmt.Errorf("failed to find OVS port %s on VM %s, err %v", vm.ifName, vm.nodeName, err)
+	}
+
+	if strings.Contains(stdout, vm.ifName) && strings.Contains(stdout, "name") {
+		return true, nil
+	}
+	return false, nil
+}
+
+func createExternalNodeCRD(data *TestData, nodeName string, ifName string, ip string) (enode *crdv1alpha1.ExternalNode, err error) {
+	testEn := &ExternalNodeSpecBuilder{}
+	testEn.SetName(nameSpace, nodeName)
+	var ipList []string
+	ipList = append(ipList, ip)
+	testEn.Spec.Interfaces = append(testEn.Spec.Interfaces, crdv1alpha1.NetworkInterface{
+		Name: ifName,
+		IPs:  ipList,
+	})
```

Comment:
I think the point of having `ExternalNodeSpecBuilder` is to encapsulate such code, otherwise it looks inconsistent to use both builder and raw code to construct an object. The builder should have a `AddInterface(name string, ips []string)` method.

Reply:
Sure.